crates/jstzd/src/docker/container.rs

```diff
@@ -1,79 +1,79 @@
-    id: String,
-    client: Arc<Docker>,
+#[derive(Default)]
+pub struct Container {
+    pub id: String,
+    client: Option<Arc<Docker>>,
+    dropped: bool,
     _private: (),
 }
 
-pub struct Container(AsyncDropper<ContainerInner>);
-impl ContainerInner {
+impl Container {
     /// Creates a new container with running `id`
     pub fn new(client: Arc<Docker>, id: String) -> Self {
         Self {
             id,
-            client,
-            // network,
+            client: Some(client),
+            dropped: false,
             _private: (),
         }
     }
 
     /// Starts the container's entrypoint
-    async fn start(&self) -> Result<()> {
-        self.client
+    pub async fn start(&self) -> Result<()> {
+        self.client()?
             .start_container::<String>(&self.id, None)
             .await?;
         Ok(())
     }
 
     /// Stop the container
-    async fn stop(&self) -> Result<()> {
-        self.client.stop_container(&self.id, None).await?;
+    pub async fn stop(&self) -> Result<()> {
+        self.client()?.stop_container(&self.id, None).await?;
         Ok(())
     }
+
+    fn client(&self) -> Result<Arc<Docker>> {
+        self.client
+            .as_ref()
+            .cloned()
+            .ok_or_else(|| anyhow::anyhow!("Client does not exist"))
+    }
 }
 
+// Stop and remove the container when dropped
 #[async_trait]
-impl AsyncDrop for ContainerInner {
+impl AsyncDrop for Container {
     async fn async_drop(&mut self) {
         self.stop()
             .await
             .unwrap_or_else(|e| error!("Error stopping container: {}", e));
-        match self.client.remove_container(&self.id, None).await {
+        let client = self.client.take().expect("Client does not exist in drop");
+        match client.remove_container(&self.id, None).await {
             Ok(_) => {}
             Err(e) => error!("Error removing container {}: {}", self.id, e),
         }
-        // TODO: https://linear.app/tezos/issue/JSTZ-111/support-for-adding-network-to-a-container
-        // Handle network drop
-    }
-}
-
-impl Container {
-    pub fn new(client: Arc<Docker>, id: String) -> Self {
-        Self(AsyncDropper::new(ContainerInner::new(client, id)))
-    }
-
-    pub fn id(&self) -> &str {
-        &self.0.inner().id
-    }
-
-    pub async fn start(&self) -> Result<()> {
-        self.0.inner().start().await
-    }
-
-    pub async fn stop(&self) -> Result<()> {
-        self.0.inner().stop().await
     }
 }
 
-#[async_trait]
-impl AsyncDrop for Container {
-    async fn async_drop(&mut self) {
-        self.0.inner_mut().async_drop().await;
+impl Drop for Container {
+    fn drop(&mut self) {
+        if !self.dropped {
+            // Prevent the copy `this` to drop again
+            self.dropped = true;
+            let mut this = std::mem::take(self);
+            // Prevent the original `self` to drop again
+            self.dropped = true;
+            TokioScope::scope_and_block(|s| {
```

Comment:
Just a note about `scope_and_block`: this only works with a multi-threaded runtime. Otherwise, the code will panic here. Sample error message:
```
thread '...' panicked at ...:                                                                                                                                                              
can call blocking only when running on the multi-threaded runtime

Reply:
I realised that this is not just for `scope_and_block` but for blocking async in general, e.g. `tokio::task::block_in_place`.

I guess we can keep that assumption for now. For unit tests this might fail since single-threaded runtime is used by default. This can be resolved with `#[tokio::test(flavor = "multi_thread")]`.